base/reshapedarray.jl

```diff
@@ -0,0 +1,107 @@
+
+# When reshaping Vector->Vector, don't wrap with a ReshapedArray
+reshape{T}(v::ReshapedArray{T,1}, dims::Tuple{Int}) = reshape(v.parent, dims[1])
+reshape(v::AbstractVector, dims::Tuple{Int}) = reshape(v, dims[1])
+function reshape(v::AbstractVector, len::Int)
+    len == length(v) || throw(DimensionMismatch("parent has $(length(v)) elements, which is incompatible with length $len"))
+    v
+end
+
+function _reshape(p::Tuple{AbstractArray,LinearSlow}, dims::Dims)
+    parent = p[1]
+    strds = front(size_strides(parent))
+    strds1 = map(s->max(1,s), strds)  # for resizing empty arrays
+    mi = map(SignedMultiplicativeInverse, strds1)
+    ReshapedArray(parent, dims, reverse(mi))
+end
+
+function _reshape(p::Tuple{AbstractArray,LinearFast}, dims::Dims)
+    parent = p[1]
+    ReshapedArray(parent, dims, ())
+end
+
+@inline size_strides(A::AbstractArray) = tail(size_strides((1,), size(A)...))
+size_strides(out::Tuple) = out
+@inline size_strides(out, s, sz...) = size_strides((out..., out[end]*s), sz...)
+
+size(A::ReshapedArray) = A.dims
+size(A::ReshapedArray, d) = d <= ndims(A) ? A.dims[d] : 1
+similar(A::ReshapedArray, eltype::Type) = similar(parent(A), eltype, size(A))
+similar(A::ReshapedArray, eltype::Type, dims...) = similar(parent(A), eltype, dims...)
+linearindexing{R<:ReshapedArrayLF}(::Type{R}) = LinearFast()
+parent(A::ReshapedArray) = A.parent
+parentindexes(A::ReshapedArray) = map(s->1:s, size(parent(A)))
+reinterpret{T}(::Type{T}, A::ReshapedArray, dims::Dims) = reinterpret(T, parent(A), dims)
+
+@inline ind2sub_rs(::Tuple{}, i::Int) = i
+@inline ind2sub_rs(strds, i) = ind2sub_rs((), strds, i-1)
+@inline ind2sub_rs(out, ::Tuple{}, ind) = (ind+1, out...)
+@inline function ind2sub_rs(out, strds, ind)
+    d, r = divrem(ind, strds[1])
+    ind2sub_rs((d+1, out...), tail(strds), r)
+end
+
+@inline getindex(A::ReshapedArrayLF, index::Int) = (@boundscheck checkbounds(A, index); @inbounds ret = parent(A)[index]; ret)
+@inline getindex(A::ReshapedArray, indexes::Int...) = (@boundscheck checkbounds(A, indexes...); _unsafe_getindex(A, indexes...))
+@inline getindex(A::ReshapedArray, index::ReshapedIndex) = (@boundscheck checkbounds(parent(A), index.parentindex); @inbounds ret = parent(A)[index.parentindex]; ret)
+
+@inline _unsafe_getindex(A::ReshapedArray, indexes::Int...) = (@inbounds ret = parent(A)[ind2sub_rs(A.mi, sub2ind(size(A), indexes...))...]; ret)
+@inline _unsafe_getindex(A::ReshapedArrayLF, indexes::Int...) = (@inbounds ret = parent(A)[sub2ind(size(A), indexes...)]; ret)
+
+@inline setindex!(A::ReshapedArrayLF, val, index::Int) = (@boundscheck checkbounds(A, index); @inbounds parent(A)[index] = val; val)
+@inline setindex!(A::ReshapedArray, val, indexes::Int...) = (@boundscheck checkbounds(A, indexes...); _unsafe_setindex!(A, val, indexes...))
+@inline setindex!(A::ReshapedArray, val, index::ReshapedIndex) = (@boundscheck checkbounds(parent(A), index.parentindex); @inbounds parent(A)[index.parentindex] = val; val)
```

Comment:
I feel like maybe "functions containing multiple statements should use the long form function definition syntax" would be a good style guideline.